overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,249 @@ import (
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdTime time.Time, affectingSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	herr := &HoldError{
+		SnapsInError: make(map[string]error),
+	}
+	now := timeNow()
+	for _, affecting := range affectingSnaps {
+		refreshed, err := lastRefreshed(st, affecting)
+		if err != nil {
+			return err
+		}
+		if now.Sub(*refreshed) > maxPostponement {
+			herr.SnapsInError[affecting] = fmt.Errorf("cannot hold the refresh of snap %q, maximum postponement time exceeded", affecting)
+			continue
+		}
+		if _, ok := gating[affecting]; !ok {
+			gating[affecting] = make(map[string]*holdInfo)
+		}
+		hold, ok := gating[affecting][gatingSnap]
+		if !ok {
+			hold = &holdInfo{
+				FirstHeld: now,
+			}
+		}
+		// determine the maximum time the snap can be held by this gating snap.
+		var maxHoldDur time.Duration
+		if affecting == gatingSnap {
+			maxHoldDur = maxPostponement
+		} else {
+			maxHoldDur = maxOtherHoldDuration
+		}
+		var holdDur time.Duration
+		// explicit hold time requested
+		if !holdTime.IsZero() {
+			holdDur = holdTime.Sub(hold.FirstHeld)
+			if holdDur > maxHoldDur {
+				herr.SnapsInError[affecting] = fmt.Errorf("requested holding time %s exceeds maximum holding time for snap %q", holdTime, affecting)
+				continue
+			}
+		} else {
+			holdDur = maxHoldDur
+		}
+		// HoldUntil is relative to the time the snap was first held, so in case
+		// of consecutive refreshes and hold attempts this cannot go beyond
+		// maximum allowed hold time.
+		hold.HoldUntil = hold.FirstHeld.Add(holdDur)
+		gating[affecting][gatingSnap] = hold
+	}
+	st.Set("refresh-gating", gating)
+	if len(herr.SnapsInError) > 0 {
+		return herr
+	}
+	return nil
+}
+
+// ProceedWithRefresh unblocks all snaps held by gatingSnap for refresh. This
+// should be called for --proceed on the gatingSnap.
+func ProceedWithRefresh(st *state.State, gatingSnap string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	if len(gating) == 0 {
+		return nil
+	}
+
+	var changed bool
+	for affectingSnap, gatingSnaps := range gating {
+		if _, ok := gatingSnaps[gatingSnap]; ok {
+			delete(gatingSnaps, gatingSnap)
+			changed = true
+		}
+		if len(gating[affectingSnap]) == 0 {
+			delete(gating, affectingSnap)
+		}
+	}
+
+	if changed {
+		st.Set("refresh-gating", gating)
+	}
+	return nil
+}
+
+// resetGating resets gating information by:
+// - removing affecting snaps whose held time expired.
+// - removing affecting snaps that are not in candidates (meaning there is no update for them anymore).
+func resetGating(st *state.State, candidates map[string]*refreshCandidate) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+
+	// optimize some edge cases
+	if len(gating) == 0 {
+		return nil
+	}
+	if len(candidates) == 0 {
```

Comment:
this is not going to be correct if we set values into refresh-gating for reasons that are the hooks (which we might at some point)